src/generators/legacy-json/utils/buildHierarchy.mjs

```diff
@@ -0,0 +1,59 @@
+/**
+ * We need the files to be in a hierarchy based off of depth, but they're
+ *  given to us flattened. So, let's fix that.
+ *
+ * Assuming that {@link entries} is in the same order as the elements are in
+ *  the markdown, we can use the entry's depth property to reassemble the
+ *  hierarchy.
+ *
+ * If depth <= 1, it's a top-level element (aka a root).
+ *
+ * If it's depth is greater than the previous entry's depth, it's a child of
+ *  the previous entry. Otherwise (if it's less than or equal to the previous
+ *  entry's depth), we need to find the entry that it was the greater than. We
+ *  can do this by just looping through entries in reverse starting at the
+ *  current index - 1.
+ *
+ * @param {Array<ApiDocMetadataEntry>} entries
+ * @returns {Array<import('../types.d.ts').HierarchizedEntry>}
+ */
+export function buildHierarchy(entries) {
+  const roots = [];
+
+  for (let i = 0; i < entries.length; i++) {
+    const entry = entries[i];
+    const currentDepth = entry.heading.depth;
+
+    if (currentDepth <= 1) {
+      // We're a top-level entry
+      roots.push(entry);
+      continue;
+    }
+
+    const previousEntry = entries[i - 1];
+
+    const previousDepth = previousEntry.heading.depth;
+    if (currentDepth > previousDepth) {
+      // We're a child of the previous one
+      if (previousEntry.hierarchyChildren === undefined) {
+        previousEntry.hierarchyChildren = [];
+      }
+
+      previousEntry.hierarchyChildren.push(entry);
+    } else {
+      // Loop to find the entry we're a child of
+      for (let j = i - 2; j >= 0; j--) {
```

Comment:
I'm afraid of this potentially becoming an infinite loop? Shouldn't there be some guardrails?

Can there be any depth higher than 6?

Reply:
> I'm afraid of this potentially becoming an infinite loop

Above we check if we're <=1, so the minimum value `j` would ever be is 0. But an infinite loop shouldn't ever happen here since the condition is that `j` is >= 0, so even at `0` it would just run once and then exit, and if `j` was some negative number somehow it wouldn't run at all.

> Can there be any depth higher than 6?

Just tried right now and it seems that the max `entry.heading.depth` can be is 5.

Reply:
> I'm afraid of this potentially becoming an infinite loop?

Since the condition is `j >= 0` I don't think this could be an infinite loop? Will still add a check though since there is an assumption being made about `i`.

> Can there be any depth higher than 6?

Can't be higher than 5 from my testing